### PR TITLE
Exempt branches matching a regular expression from branch protection

### LINKED ISSUE #12954: exempt branches with a regex from branch protection
**What would you like to be added**:
Allow for the exemption of certain branches from branch protection, based on matching a regex.

**Why is this needed**:
This would allow for easy deletion of branch accidentally created via GitHub's edit UI, as well as branches created by bots (such as GitHub's dependabot)

### PR DESCRIPTION
**Pull request by adshmh**
Closes #12954

### CODE REVIEW

File: prow/config/branch_protection.go [hunk 1]
```diff
@@ -194,7 +200,8 @@ func (o Org) GetRepo(name string) *Repo {
 // Repo holds protection policy overrides for all branches in a repo, as well as specific branch overrides.
 type Repo struct {
 	Policy
-	Branches map[string]Branch `json:"branches,omitempty"`
+	Branches         map[string]Branch `json:"branches,omitempty"`
+	branchExclusions *regexp.Regexp
 }
 
 // GetBranch returns the branch config after merging in any repo policies.
```
Comment (stevekuznetsov) [inline prow/config/branch_protection.go]: I do not think `GetBranch` should be returning `nil` -- we should not make the call to `GetBranch` in the first place.

Comment (adshmh) [reply prow/config/branch_protection.go]: Thank you for the review. Do you mean to move the logic for handling branch exclusion into the `protect` command itself instead and let `config` objects provide policies for branches based on the org/repo/branch hierarchy?

Comment (stevekuznetsov) [reply prow/config/branch_protection.go]: Yes. When resolving the list of branches we need to know protection rules for, handle the exclusions there so we're not introducing `nil` checks everywhee

Comment (adshmh) [reply prow/config/branch_protection.go]: Updated the PR. Not sure why the file `prow/monitoring/BUILD.bazel` gets updated by running `hack/update-bazel.sh`.

Comment (stevekuznetsov) [reply prow/config/branch_protection.go]: That file had some issues... should be fixed if you rebase on `master`

**Comment by adshmh** [comment]
Test failure was due to `EnforceAdmins` field now always being set in branch protection requests. Fixed.

**Comment by cblecker** [comment]
This is cool! Thanks for this @adshmh ! 🚀

**Pull request merged**